File: dotnet/Lexy.Poc/Lexy.Poc.Core/Compiler/ExecutionEnvironment.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using Lexy.Poc.Core.Language;
using Lexy.Poc.Core.Parser;
using Lexy.Poc.Core.RunTime;
using Lexy.Poc.Core.Transcribe;

namespace Lexy.Poc.Core.Compiler
{
    public class ExecutionEnvironment : IExecutionEnvironment
    {
        private readonly IList<GeneratedClass> generatedTypes = new List<GeneratedClass>();
        private readonly IDictionary<string, ExecutableFunction> executables = new Dictionary<string, ExecutableFunction>();
        private readonly IDictionary<string, Type> enums = new Dictionary<string, Type>();
        private readonly IDictionary<string, Type> tables = new Dictionary<string, Type>();

        private readonly IExecutionContext executionContext;

        public ExecutionEnvironment(IExecutionContext executionContext)
        {
            this.executionContext = executionContext ?? throw new ArgumentNullException(nameof(executionContext));
        }

        public void CreateExecutables(Assembly assembly)
        {
            if (assembly == null) throw new ArgumentNullException(nameof(assembly));

            foreach (var generatedClass in generatedTypes)
            {
                switch (generatedClass.Component)
                {
                    case Function _:
                    {
                        var instance = assembly.CreateInstance(generatedClass.FullClassName);
                        var executable = new ExecutableFunction(instance, executionContext);

                        executables.Add(generatedClass.Component.ComponentName, executable);
                        break;
                    }
                    case EnumDefinition _:
                    {
                        var enumType = assembly.GetType(generatedClass.FullClassName);
                        enums.Add(generatedClass.Component.ComponentName, enumType);
                        break;
                    }
                    case Table _:
                    {
                        var table = assembly.GetType(generatedClass.FullClassName);
                        tables.Add(generatedClass.Component.ComponentName, table);
                        break;
                    }
                    default:
                        throw new InvalidOperationException("Unknown generated type: " + generatedClass.Component.GetType());
                }
            }
        }

        public void AddType(GeneratedClass generatedType)
        {
            generatedTypes.Add(generatedType);
        }

        public CompilerResult Result()
        {
            return new CompilerResult(executables, enums);
        }
    }
}

File: typescipt/existing/src/compiler/executionEnvironment.ts









namespace Lexy.Compiler.Compiler;

public class ExecutionEnvironment : IExecutionEnvironment
{
   private readonly IDictionary<string, Type> enums = new Dictionary<string, Type>();
   private readonly IDictionary<string, ExecutableFunction> executables = new Dictionary<string, ExecutableFunction>();

   private readonly IExecutionContext executionContext;
   private readonly IList<GeneratedClass> generatedTypes = new List<GeneratedClass>();
   private readonly IDictionary<string, Type> tables = new Dictionary<string, Type>();
   private readonly IDictionary<string, Type> types = new Dictionary<string, Type>();

   public ExecutionEnvironment(IExecutionContext executionContext)
   {
     this.executionContext = executionContext ?? throw new ArgumentNullException(nameof(executionContext));
   }

   public void CreateExecutables(Assembly assembly)
   {
     if (assembly = null) throw new ArgumentNullException(nameof(assembly));

     foreach (var generatedClass in generatedTypes) CreateExecutable(assembly, generatedClass);
   }

   public void AddType(GeneratedClass generatedType)
   {
     generatedTypes.Add(generatedType);
   }

   public CompilerResult Result()
   {
     return new CompilerResult(executables, enums);
   }

   private void CreateExecutable(Assembly assembly, GeneratedClass generatedClass)
   {
     switch (generatedClass.Node)
     {
       case Function _:
       {
         var instanceType = assembly.GetType(generatedClass.FullClassName);
         var executable = new ExecutableFunction(instanceType, executionContext);

         executables.Add(generatedClass.Node.NodeName, executable);
         break;
       }
       case EnumDefinition _:
       {
         CreateExecutable(assembly, generatedClass, enums);
         break;
       }
       case Table _:
       {
         CreateExecutable(assembly, generatedClass, tables);
         break;
       }
       case TypeDefinition _:
       {
         CreateExecutable(assembly, generatedClass, types);
         break;
       }
       default:
       {
         throw new InvalidOperationException("Unknown generated type: " + generatedClass.Node.GetType());
       }
     }
   }

   private void CreateExecutable(Assembly assembly, GeneratedClass generatedClass,
     IDictionary<string, Type> dictionary)
   {
     var instanceType = assembly.GetType(generatedClass.FullClassName);

     dictionary.Add(generatedClass.Node.NodeName, instanceType);
   }
}
